arroyo/processing/strategies/streaming/collect.py

```diff
@@ -97,54 +98,106 @@ def __init__(
 
     def close(self) -> None:
         self.__closed = True
 
-        if self.__batch is not None:
-            logger.debug("Closing %r...", self.__batch)
-            self.__batch.close()
+        if self.batch is not None:
+            logger.debug("Closing %r...", self.batch)
+            self.batch.close()
 
     def terminate(self) -> None:
         self.__closed = True
 
-        if self.__batch is not None:
-            self.__batch.terminate()
+        if self.batch is not None:
+            self.batch.terminate()
 
     def join(self, timeout: Optional[float] = None) -> None:
-        if self.__batch is not None:
-            self.__batch.join(timeout)
-            logger.info("Completed processing %r.", self.__batch)
-            self.__batch = None
+        if self.batch is not None:
+            self.batch.join(timeout)
+            logger.info("Completed processing %r.", self.batch)
+            self.batch = None
+
+
+class ParallelCollectStep(CollectStep[TPayload]):
+    """
+    ParallelCollectStep is similar to CollectStep except it allows the closing and reset of the
+    batch to happen in a threadpool. What this allows for is the next batch to start getting
+    filled in while the previous batch is still being processed.
+
+    The threadpool will have only 1 worker since we want to perform writes to clickhouse sequentially
+    so that kafka offsets are written in order.
+    """
+
+    def __init__(
+        self,
+        step_factory: Callable[[], ProcessingStep[TPayload]],
+        commit_function: Callable[[Mapping[Partition, Position]], None],
+        max_batch_size: int,
+        max_batch_time: float,
+    ):
+        super().__init__(step_factory, commit_function, max_batch_size, max_batch_time)
+        self.__threadpool = ThreadPoolExecutor(max_workers=1)
+        self.future: Optional[Future[None]] = None
+
+    def close_and_reset_batch(self) -> None:
+        """
+        Closes the current batch in an asynchronous manner. Waits for previous work to be completed before proceeding
+        the next one. We can provide the existing batch to the threadpool since the collector is going to make a
+        new batch.
+        """
+        if self.future:
+            # If any exceptions are raised they should get bubbled up.
+            self.future.result(timeout=5)
+
+        self.future = self.__threadpool.submit(self.__finish_batch, batch=self.batch)
+        self.batch = None
+
+    @staticmethod
+    def __finish_batch(batch: Batch[TPayload]) -> None:
+        assert batch is not None
+
+        batch.close()
+        batch.join()
+        logger.info("Completed processing %r.", batch)
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        # We should finish the previous batch before proceeding to the finish the existing one.
+        if self.future is not None:
+            self.future.result(timeout)
+
+        self.__threadpool.shutdown()
+
+        super().join(timeout)
```

Comment:
Not a huge deal but I think it would be more correct to pass the remaining time here and not the entire timeout.